modules/dnn/src/layers/pooling_layer.cpp

```diff
@@ -1089,8 +1216,125 @@ class PoolingLayerImpl CV_FINAL : public PoolingLayer
                             const float* first = srcData + xstart;
                             const float* last = srcData + xend;
-                            float sum_val = std::accumulate(first, last, 0.f);
-                            dstData[x0] = sum_val*inv_kernel_area;
+                            float sum_val = 0.f;
+                            switch (poolingType) {
+                                case AVE: case SUM:
+                                {
+                                    sum_val = std::accumulate(first, last, 0.f);
+                                    dstData[x0] = sum_val*inv_kernel_area;
+                                    break;
+                                }
+                                case L1:
+                                {
+                                    sum_val = std::accumulate(first, last, 0.f,
+                                    [](float a, float b)
+                                    {
+                                        return a + std::abs(b);
+                                    });
+                                    dstData[x0] = sum_val;
+                                    break;
+                                }
+                                case L2: case SUM_SQUARE:
+                                {
+                                    sum_val = std::accumulate(first, last, 0.f,
+                                    [](float a, float b)
+                                    {
+                                        return a + b * b;
+                                    });
+                                    dstData[x0] = poolingType == L2 ? std::sqrt(sum_val) : sum_val;
+                                    break;
+                                }
+                                case PROD:
+                                {
+                                    sum_val = std::accumulate(first, last, 1.0f, std::multiplies<float>());
+                                    dstData[x0] = sum_val;
+                                    break;
+                                }
+                                default:
+                                    CV_Error(Error::StsNotImplemented, "Not implemented");
+                                    break;
+                            }
+                        }
+                        else
+                        {
+                            float sum_val = poolingType == PROD ? 1.0f : 0.f;
+                            for (int d = dstart; d < dend; ++d) {
+                                for (int y = ystart; y < yend; ++y) {
+                                    for (int x = xstart; x < xend; ++x) {
+                                        const int index = d * inp_width * inp_height + y * inp_width + x;
+                                        float val = srcData[index];
+
+                                        switch (poolingType) {
+                                            case AVE: case SUM:
```

Comment:
Having a `switch` in triple nested `for` loops are terrible from performance perspective.

We need dedicated "processing runners" for each type. E.g., using templates with functors.

---

We should not have performance regressions during implementing features.

Reply:
@alalek, I think, the compiler should be able to optimize it since we're not changing poolingType, but you're right, dedicated runners make more sense.